Source/JavaScriptCore/CMakeLists.txt: ensure reproducibility of __TIMESTAMP__

kanavin opened the pull request:
#### 8e2834324073319c620af6ff1929827b8b76fef9
<pre>
Source/JavaScriptCore/CMakeLists.txt: ensure reproducibility of __TIMESTAMP__
<a href="https://bugs.webkit.org/show_bug.cgi?id=258517">https://bugs.webkit.org/show_bug.cgi?id=258517</a>

Reviewed by Michael Catanzaro.

__TIMESTAMP__ refers to mtime of the file that contains it, which is unstable
and breaks binary reproducibility when the file is generated at build time. To ensure
this does not happen, mtime should be set from the original file.

* Source/JavaScriptCore/CMakeLists.txt:

Canonical link: <a href="https://commits.webkit.org/266322@main">https://commits.webkit.org/266322@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d003b03a2b68c6d5671163274a542fd2216251f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13526 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13840 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14171 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15263 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12865 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16348 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13917 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15548 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13693 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14338 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11450 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15967 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11625 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12205 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19252 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/11526 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12701 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12373 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15586 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/12807 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12887 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10770 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/13570 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12155 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3544 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3293 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16483 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/13955 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12730 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3345 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/JavaScriptCore/CMakeLists.txt:
```diff
@@ -159,6 +159,7 @@ add_custom_command(
     OUTPUT ${JavaScriptCore_DERIVED_SOURCES_DIR}/JSCBytecodeCacheVersion.cpp
     MAIN_DEPENDENCY ${JAVASCRIPTCORE_DIR}/runtime/JSCBytecodeCacheVersion.cpp.in
     COMMAND ${PERL_EXECUTABLE} -pe s/CACHED_TYPES_CKSUM/__TIMESTAMP__/ ${JAVASCRIPTCORE_DIR}/runtime/JSCBytecodeCacheVersion.cpp.in > ${JavaScriptCore_DERIVED_SOURCES_DIR}/JSCBytecodeCacheVersion.cpp
+    COMMAND touch -r ${JAVASCRIPTCORE_DIR}/runtime/JSCBytecodeCacheVersion.cpp.in ${JavaScriptCore_DERIVED_SOURCES_DIR}/JSCBytecodeCacheVersion.cpp
```
This is actually a pretty good idea!

Except you can't use `touch` here, because it doesn't exist on Windows and this is a cross-platform file (which is why the Windows CI failed). I looked for an equivalent CMake command and found [this](https://cmake.org/cmake/help/latest/command/file.html#touch), but it's not good enough for what you're trying to do.

We'll have to think of something else.

emw-apple replied:
I'm not sure whether you could support it in all the environments you need to build in, but one thing we considered in https://github.com/WebKit/WebKit/commit/adb96155be5e75707c6d831c3329b72b0afcc2b1 was using a string from the git repo's state (e.g. `git describe --dirty`).

mcatanzaro replied:
That would be OK for developers building from git, but won't work for tarball builds.

I wonder if CMake has a way to check if `touch` is installed. (Can't use `which` because, again, that's a Windows thing.)

emw-apple replied:
Hang on, can we just use [`cmake -E touch`](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-E-arg-touch)?

kanavin replied:
> Hang on, can we just use [`cmake -E touch`](https://cmake.org/cmake/help/latest/manual/cmake.1.html#cmdoption-cmake-E-arg-touch)?

We need touch's -r option that sets times from a reference file, which cmake -E touch does not support.

mcatanzaro replied:
Another option: if there's a way to test whether `touch` is installed, then we could have reproducible builds on systems that do have `touch`.

kanavin replied:
> Another option: if there's a way to test whether `touch` is installed, then we could have reproducible builds on systems that do have `touch`.

I've adjusted the commit to do that, please review.

mcatanzaro reviewed (changes requested):
.

kanavin committed:
Source/JavaScriptCore/CMakeLists.txt: ensure reproducibility of __TIMESTAMP__
https://bugs.webkit.org/show_bug.cgi?id=258517

Reviewed by Michael Catanzaro.

__TIMESTAMP__ refers to mtime of the file that contains it, which is unstable
and breaks binary reproducibility when the file is generated at build time. To ensure
this does not happen, mtime should be set from the original file.

* Source/JavaScriptCore/CMakeLists.txt:

Canonical link: https://commits.webkit.org/266322@main
